src/kmeans.jl

```diff
@@ -1,91 +1,132 @@
 # K-means algorithm
 
-#### Interface
-
-mutable struct KmeansResult{T<:Real} <: ClusteringResult
-    centers::Matrix{T}         # cluster centers (d x k)
-    assignments::Vector{Int}   # assignments (n)
-    costs::Vector{T}           # costs of the resultant assignments (n)
-    counts::Vector{Int}        # number of samples assigned to each cluster (k)
-    cweights::Vector{Float64}  # cluster weights (k)
-    totalcost::Float64         # total cost (i.e. objective) (k)
-    iterations::Int            # number of elapsed iterations
-    converged::Bool            # whether the procedure converged
+####
+#### Result object
+####
+
+# NOTE REFACT: there's no need for this to be a mutable struct.
+
+struct KmeansResult{T<:AbstractFloat} <: ClusteringResult
+    centers::Matrix{T}        # cluster centers (d x k)
+    assignments::Vector{Int}  # assignments (n)
+    costs::Vector{T}          # cost of the assignments (n)
+    counts::Vector{Int}       # number of samples assigned to each cluster (k)
+    cweights::Vector{T}       # cluster weights (k)
+    totalcost::T              # total cost (i.e. objective)
+    iterations::Int           # number of elapsed iterations
+    converged::Bool           # whether the procedure converged
 end
 
 const _kmeans_default_init = :kmpp
 const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::AbstractMatrix{T}, centers::AbstractMatrix{T};
+
+####
+#### Exported kmeans and kmeans! functions
+####
+
+# Case when TX is a Real but not a AbstractFloat
+function kmeans!(X::AbstractMatrix{<:Union{Integer, Rational, AbstractIrrational}},
+                 centers::AbstractMatrix{<:Real}; args...)
+    # NOTE REFACT: a warning could be issued here
+    kmeans!(convert(Matrix{Float64}, X), centers; args...)
+end
+
+# case when TX<:AbstractFloat
+function kmeans!(X::AbstractMatrix{TX}, centers::AbstractMatrix{TC};
                  weights::Union{Nothing, AbstractVector{<:Real}}=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
+                 maxiter::Int=_kmeans_default_maxiter,
                  tol::Real=_kmeans_default_tol,
                  display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:Real
+                 distance::SemiMetric=SqEuclidean()
+                 ) where {TX<:AbstractFloat,TC<:Real}
```

Comment:
Actually, `TC <: AbstractFloat` would make more sense here.
`MethodError` indicating that integral `centers` are not supported would be much clearer than `InexactError` in `update_centers!()`

Reply:
yes this is done in 992c8bb 😛